### ISSUE TITLE
make fish_update_completions compatible with m(an)doc

### ISSUE THREAD

**Issue by gour** [fish-shell/fish-shell#2194]
Hello,

installed 2.2.0 on VoidLinux which uses m(an)doc instead of man-db and fish_update_completions function crashes due to missing **manpath** executable:

```
Traceback (most recent call last):
  File "/usr/share/fish/tools/create_manpage_completions.py", line 963, in <module>
    file_paths.extend(get_paths_from_manpath())
  File "/usr/share/fish/tools/create_manpage_completions.py", line 894, in get_paths_from_manpath
    proc = subprocess.Popen(['manpath'], stdout=subprocess.PIPE)
  File "/usr/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1335, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```

so it would be nice if _create_manpage_completions.py_ script could be made compatible with the systems/OS-es which use m(an)doc?

**Comment by faho** [comment]
What's the equivalent call? How do you get the paths that mandoc's man would search? Or is it just a case of "Either $MANPATH or /usr/share/man" (i.e. hardcoded)?

I tried looking through mandoc's man pages, but could not find any such thing.

**Comment by gour** [comment]
> What's the equivalent call? How do you get the paths that mandoc's man would search? Or is it just a case of "Either $MANPATH or /usr/share/man" (i.e. hardcoded)?

Probably the latter, although I'm not much familiar with mandoc.

> I tried looking through mandoc's man pages, but could not find any such thing.

I did ask in #openbsd yesterday and some guys found out that e.g. on FreeBSD manpath is shell script. However, I'm not sure whether hardcoding path to e.g. /usr/share/man would be general-enough solution for mandoc-based systems, but I strongly believe that mandoc should be considered as option.

Will try to investigate further...

**Comment by faho** [comment]
I checked a bit further, found [this pdf](http://www.openbsd.org/papers/bsdcan15-mandoc.pdf) that talks about /etc/man.conf (which doesn't have a corresponding manpage or any other kind of documentation I can find), which can contain "manpath VALUE".

So it seems like the precedence is:

- Environment variable $MANPATH
- /etc/man.conf's "manpath"
- the built-in default - this _is_ a build-time variable but I can't find a way to access it at runtime

So it seems like the best we can currently do is something like (UNTESTED, RUSTY PYTHON, CHAINSAWS!):

    if os.getenv("MANPATH"):
        manpath = os.getenv("MANPATH")
    else:
        # Read /etc/man.conf for manpath
    try:
        proc = subprocess.Popen(['manpath'], stdout=subprocess.PIPE)
        manpath, err_data = proc.communicate()
    except OSError:
        manpath="/usr/share/man"

**Comment by gour** [comment]
It looks good - I just took a look at my /etc/man.conf and see stuff like:

```
# The whatis/apropos database.
_whatdb         /usr/share/man/mandoc.db
_whatdb         /usr/local/share/man/mandoc.db
[...]
# Sections and their directories.
# All paths ending in '/' are the equivalent of entries specifying that
# directory with all of the subdirectories listed for the keyword _subdir.

# default
_default        /usr/{share,local/share}/man/

# Other sections that represent complete man subdirectories.
local           /usr/local/share/man/
```

**Comment by zanchey** [comment]
Brace expansion in a configuration file? That's amazing.

**Comment by seisatsu** [comment]
I'm on Void Linux and this is still an issue in version 2.2.0. faho's patch seems to work.

**Comment by Shados** [comment]
At the very least, making `fish-shell/share/tools/create_manpage_completions.py` check $MANPATH before resorting to running `manpath` would bring its behaviour in-line with that of `fish-shell/share/functions/man.fish`. Might also be worthwhile checking `man --path` if `manpath` fails, as you don't know what implementation of `man` the OS is using.

**Comment by pickfire** [comment]
I just tried void and found out that `fish_update_completions` doesn't work. Here's the man.conf to aid in development:
```
#       $OpenBSD: man.conf,v 1.20 2014/02/24 20:11:53 espie Exp $

# Sheer, raging paranoia...
_version        BSD.2

# The whatis/apropos database.
_whatdb         /usr/share/man/mandoc.db
_whatdb         /usr/local/share/man/mandoc.db

# Subdirectories for paths ending in '/', IN SEARCH ORDER.
_subdir         {cat,man}0p {cat,man}1 {cat,man}1p {cat,man}8 {cat,man}6 {cat,man}2 {cat,man}3 {cat,man}5 {cat,man}7 {cat,man}4 {cat,man}9 {cat,man}3p {cat,man}3f {cat,man}n

# Files typed by suffix and their commands.
_suffix         .0
_build          .0.gz                   /usr/bin/zcat %s
_build          .[0-9n]                 /usr/bin/mandoc -Tutf8 %s
_build          .[0-9n].gz              /usr/bin/zcat %s | /usr/bin/mandoc -Tutf8
_build          .[0-9][a-z]             /usr/bin/mandoc -Tutf8 %s
_build          .[0-9][a-z].gz          /usr/bin/zcat %s | /usr/bin/mandoc -Tutf8
_build          .[0-9][a-z][a-z]        /usr/bin/mandoc -Tutf8 %s
_build          .[0-9][a-z][a-z].gz     /usr/bin/zcat %s | /usr/bin/mandoc -Tutf8
_build          .[0-9][a-z][a-z][a-z]   /usr/bin/mandoc -Tutf8 %s
_build          .[0-9][a-z][a-z][a-z].gz        /usr/bin/zcat %s | /usr/bin/mandoc -Tutf8
_build          .tbl                    /usr/bin/mandoc -Tutf8 %s
_build          .tbl.gz                 /usr/bin/zcat %s | /usr/bin/mandoc -Tutf8

# Sections and their directories.
# All paths ending in '/' are the equivalent of entries specifying that
# directory with all of the subdirectories listed for the keyword _subdir.

# default
_default        /usr/{share,local/share}/man/

# Other sections that represent complete man subdirectories.
local           /usr/local/share/man/

# Specific section/directory combinations.
0p              /usr/{share,local/share}/man/{cat,man}0p
1               /usr/{share,local/share}/man/{cat,man}1
1p              /usr/{share,local/share}/man/{cat,man}1p
2               /usr/{share,local/share}/man/{cat,man}2
3               /usr/{share,local/share}/man/{cat,man}3
3F              /usr/{share,local/share}/man/{cat,man}3f
3f              /usr/{share,local/share}/man/{cat,man}3f
3P              /usr/{share,local/share}/man/{cat,man}3p
3p              /usr/{share,local/share}/man/{cat,man}3p
3x              /usr/{share,local/share}/man/{cat,man}3
3am             /usr/{share,local/share}/man/{cat,man}3
4               /usr/{share,local/share}/man/{cat,man}4
5               /usr/{share,local/share}/man/{cat,man}5
6               /usr/{share,local/share}/man/{cat,man}6
7               /usr/{share,local/share}/man/{cat,man}7
8               /usr/{share,local/share}/man/{cat,man}8
9               /usr/{share,local/share}/man/{cat,man}9
n               /usr/{share,local/share}/man/{cat,man}n
```

**Comment by faho** [comment]
Please test [my manpath branch](https://github.com/faho/fish-shell/tree/manpath).

**Comment by pickfire** [comment]
@faho Tested, does not work here on `Linux void 4.8.15_1 #1 SMP PREEMPT Thu Dec 15 23:12:48 UTC 2016 x86_64 GNU/Linux`.
```
> fish -v
fish, version 2.1.1-3126-ge1496aab
> echo $version
2.1.1-3126-ge1496aab
> fish_update_completions
man: unrecognized option: -
man: unrecognized option: p
man: unrecognized option: t
usage: man [-acfhklw] [-C file] [-I os=name] [-K encoding] [-M path] [-m path]
           [-O option=value] [-S subsection] [-s section] [-T output] [-W level]
           [section] name ...
No paths specified
```

**Comment by faho** [comment]
Welp, I don't know what I expected there. Of course it's not going to work because calling `man --path` actually _works_ - `man` exists. But it _fails_ - returns non-zero, and prints errors.

Now I need to look up how that works in python - I've never actually done anything with subprocess before.

**Comment by faho** [comment]
@pickfire: Please test again with 09439ac8d4e9ff23938471b66848143d596255d.

**Comment by pickfire** [comment]
@faho From what I look at the code, I think probably it will work. I have never used Popen myself. Will be testing it soon when I on Void.

**Comment by pickfire** [comment]
@faho, tested and it works. Output:
```
> fish_update_completions
Unable to get the manpath, falling back to /usr/share/man:/usr/local/share/man. Please
set $MANPATH if that is not correct.Parsing man pages and writing completions to /home/ivan/.local/share/fish/generated_completions/
  1527 / 1527 : zzz.8
```
It seems sad that everytime I need to do the following in order to build fish:

    sudo xbps-install base-devel ncurses-devel python

Well, python and python3 is not installed here, need to manually install them to use fish_update_completions.

**Comment by pickfire** [comment]
But there seems to have some problems during `sudo make install`:
```
  FISH_BUILD_VERSION = '2.1.1-3127-g209439ac'
  DESTDIR = ''
  localedir = '/usr/local/share/locale'
  prefix = '/usr/local'
  bindir = '/usr/local/bin'
  sysconfdir = '/usr/local/etc'
  datadir = '/usr/local/share'
  docdir = '/usr/local/share/doc/fish'
  mandir = '/usr/local/share/man'

Uninstalling fish from configured $prefix: /usr/local

Deleting programs: [ fish fish_indent fish_key_reader ] in /usr/local/bin
Deleting configuration: /usr/local/etc/fish/*

In 5 seconds, all data (includes functions, completions, tools) in
$       /usr/local/share/fish will be deleted!

$If you put things there, stop now! $5 $4 $3 $2 $1 ...
... deleting /usr/local/share/fish/*

Deleting documentation: /usr/local/share/doc/fish/*

Deleting pkgconfig file: /usr/local/share/pkgconfig/fish.pc

$Deleting [ share/man/man1/fish share/man/man1/fish_indent share/man/man1/fish_key_reader $] in /usr/local/share/man/man1
/bin/sh: 2: Syntax error: end of file unexpected (expecting "done")
make: [Makefile:771: uninstall] Error 2 (ignored)
done;
/bin/sh: 1: Syntax error: "done" unexpected
make: *** [Makefile:772: uninstall] Error 2
```

**Comment by faho** [comment]
>tested and it works

Thanks!

> Output:

Okay, that needs another newline.

>But there seems to have some false error during sudo make install:

@pickfire: Since I'm reasonably sure that's not caused by this PR, please open a new issue about it.

>Well, python and python3 is not installed here, need to manually install them to use fish_update_completions.

Yes, we use some python scripts. Python is rather well suited to this kind of work and also available everywhere.  Most systems will have something depending on it already, and if you don't you can either install it or not use the features depending on it. None of them (currently it's the completions generator and fish_config) are critical. We should of course handle a missing python gracefully.

>Hope it gets re-written in C++ or C

Sorry, that's not going to happen.

>I hope this is in the next-minor milestone.

My plan is to merge it this week, so it should be included in the release after 2.5.0. Maybe even 2.5.0. @zanchey?
